### PR TITLE
feature: Add multi-node connection pool

### LINKED ISSUE #57: [ENHANCEMENT] Implement SniffingConnectionPool
**Is your feature request related to a problem? Please describe.**
The client can work against an Elasticsearch cluster behind a single endpoint e.g. a single node cluster, a cluster behind a proxy/load balancer. It is also useful however to be able to seed a client with a collection of endpoints, where each endpoint connects to an Elasticsearch node. In doing so, it allows for future features such as retries/failover, executing  requests on a node matching a predicate e.g. data nodes only, etc. Since Elasticsearch clusters can change in size, the client should be able to reseed itsefl with the nodes (endpoints) available within the cluster. This is commonly referred to across the other Elasticsearch clients as _sniffing_.

**Describe the solution you'd like**
A `SniffingConnectionPool` that implements `ConnectionPool` that can

1. be seeded with an initial collection of `Url`
2. returns the next `Connection` from the pool when `next()` is called. In the initial implementation, this can simply iterate over connections, but in future, can mark connections dead/alive, weight them by aliveness, etc. to determine the next connection to return
3. can periodically reseed the connections by _sniffing_ the cluster

**Additional context**

An example implementation of [`SniffingConnectionPool` from the .NET client](https://github.com/elastic/elasticsearch-net/blob/master/src/Elasticsearch.Net/ConnectionPool/SniffingConnectionPool.cs) and the [behaviour](https://www.elastic.co/guide/en/elasticsearch/client/net-api/current/sniffing-behaviour.html)

### PR DESCRIPTION
**Pull request by tommilligan**
Closes #57
Based on/supersedes #139 by @srleyva

Initially reviewed by @russcam and @swallez 

---

This PR rebases #139 on top of current master, updates the added dependencies to current versions, and fixes clippy/deprecation warnings present in the diff.

I have maintained the original commit history for ease of picking up review from the existing PR - I believe everything after and including `Add regex parsing bound_address to URL` is unreviewed.

I have a couple of thoughts of potential improvements, but would like to hear from others before I sink any more time into this.

### CODE REVIEW

File: elasticsearch/src/http/transport.rs [hunk 1]
```diff
@@ -474,6 +519,102 @@ impl Transport {
...
+                &connection,
+                Method::Get,
+                "_nodes/http?filter_path=nodes.*.http",
+                headers.clone(),
+                None::<&Q>,
+                None::<B>,
+                timeout,
+            )?;
+
+            spawn(move || {
```
Comment (mfelsche) [inline elasticsearch/src/http/transport.rs:577]: if we require tokio to be around anyways, wouldn't it be better (in terms of resource usage etc.) to spawn a tokio task for the reseeding?

File: elasticsearch/src/http/transport.rs [hunk 2]
```diff
@@ -474,6 +519,102 @@ impl Transport {
...
         }
+        Ok(request_builder)
+    }
+
+    fn parse_to_url(address: &str, scheme: &str) -> Result<Url, Error> {
+        if address.is_empty() {
+            return Err(crate::error::lib("Bound Address is empty"));
+        }
+
+        let matches = ADDRESS_REGEX
```
Comment (mfelsche) [inline elasticsearch/src/http/transport.rs:530]: I am a little uncomfortable using regex based parsing here. Wouldn't it be easier to check the `address` str for a valid scheme prefix? That would be `http://` `https://` and if it doesn't have one, prepend the given `scheme`, parse the url and then use `Url::set_scheme`?

**Comment by Licenser** [comment]
From the perspective of a user of the library, I'm a bit concerned over the number of `unwrap()` calls in the added logic. Would it be possible to replace them with some kind of handle error?

A library that uwraps/panics is really hard to deal with when it comes to providing helpful errors to users, using Result and some kind of Error allows the application to provide detail, hints, and additional help to a user as to what went wrong also it can then handle a failure gracefully - after all a hard stop of an application is not always the desired behavior on a failure in a library.

**Review by swallez — APPROVED** [review]
I have rebased on the latest `main` along with some fine-tuning (simplify node address parsing, and run reseed as a tokio task instead of a thread with a new runtime)

**Pull request merged**